src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,416 @@
+      /* root is NULL if the file is empty */
+      if (result != NULL) {
+        if (JSON_NODE_HOLDS_OBJECT (result))
+          result = json_node_copy (result);
+        else
+          {
+            g_warning ("%s does not contain a JSON object, ignoring", path);
+            result = NULL;
+          }
+      }
+    }
+  else
+    {
+      if (error->code != G_FILE_ERROR_NOENT)
+        g_warning("%s has invalid JSON: %s", path, error->message);
+      g_error_free (error);
+    }
+
+  g_object_unref (parser);
+  return result;
+}
+
+static void
+merge_config (JsonObject *machines, JsonObject *delta, const char *path)
+{
+  GList *hosts = json_object_get_members (delta);
+  for (GList *i = g_list_first (hosts); i; i = g_list_next (i))
+    {
+      const char *hostname = i->data;
+      JsonNode *delta_props = json_object_get_member (delta, hostname);
+
+      if (!JSON_NODE_HOLDS_OBJECT (delta_props))
+        {
+          g_warning ("%s: host name definition %s does not contain a JSON object, ignoring", path, hostname);
+          continue;
+        }
+
+      /* merge delta properties info existing machines host */
+      if (!json_object_has_member (machines, hostname))
+        json_object_set_member (machines, hostname, new_object_node ());
+      JsonObject *machine_props = json_object_get_object_member (machines, hostname);
+
+      g_debug("%s: merging updates to host name %s", path, hostname);
+      GList *proplist = json_object_get_members (json_node_get_object (delta_props));
+      for (GList *p = g_list_first (proplist); p; p = g_list_next (p))
+        {
+          const char *propname = p->data;
+          JsonNode *prop_node = json_object_get_member (json_node_get_object (delta_props), propname);
+
+          if (!JSON_NODE_HOLDS_VALUE (prop_node))
+            {
+              g_warning ("%s: host name definition %s: property %s does not contain a simple value, ignoring", path, hostname, propname);
+              continue;
+            }
+
+          g_debug ("%s:  host name %s: merging property %s", path, hostname, propname);
+          json_object_set_member (machine_props, propname, json_node_copy (prop_node));
+        }
+
+      g_list_free (proplist);
+    }
+
+  g_list_free (hosts);
+}
+
+static JsonNode*
+read_machines_json (void)
+{
+  gchar *glob_str;
+  glob_t conf_glob = { .gl_offs = 1 };
+  int res;
+  JsonNode *machines = NULL;
+
+  /* start with an empty object */
+  machines = new_object_node ();
+
+  /* find json config files */
+  glob_str = g_build_filename (machines_json_dir(), "*.json", NULL);
+  res = glob (glob_str, GLOB_DOOFFS, glob_err_func, &conf_glob);
+  if (G_UNLIKELY (res != 0 && res != GLOB_NOMATCH))
+    g_error ("glob %s failed with return code %i", glob_str, res);
+  /* also read /var/lib/cockpit/machines.json for backwards compat */
+  conf_glob.gl_pathv[0] = g_build_filename (g_getenv ("COCKPIT_DATA_DIR") ?: "/var/lib/cockpit", "machines.json", NULL);
```

Comment:
Migrating the file is part of https://github.com/cockpit-project/cockpit/wiki/Config-format-for-known-machines-and-ssh-keys, this PR just does not implement this (yet). I wasn't completely sure if we can just do that, but I suppose we don't need to be concerned about downgrading -bridge after the migration?

However, it might happen that the migration fails because `/etc/cockpit/machines.d/` cannot be created or written to for some reason. So we need to keep the code that reads `/var/` for some time at least.

This seemed hairy enough to be in a separate PR (this doesn't address known_hosts either yet), but if you prefer I'll fold it in here.